### PR TITLE
feat(temporal-filter): support more now expressions in temporal filter pattern

### LINKED ISSUE #17630: allow conditions like `ts < expr(now())` in temporal filter

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support complex expressions like `ts + interval '1 hour' > date_trunc('day', ('2024-07-18 00:00:00+00:00'::timestamptz - ('2024-07-18 00:00:00+00:00'::timestamptz - now()))`.

Fixes #17630.

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/output/temporal_filter.yaml [hunk 1]
```diff
@@ -460,3 +460,29 @@
...
+    select * from t where ts + interval '1 hour' > date_trunc('day', ('2024-07-18 00:00:00+00:00'::timestamptz - ('2024-07-18 00:00:00+00:00'::timestamptz - now())));
+  stream_plan: |-
+    StreamMaterialize { columns: [ts, a, t._row_id(hidden)], stream_key: [t._row_id], pk_columns: [t._row_id], pk_conflict: NoCheck }
+    └─StreamProject { exprs: [t.ts, t.a, t._row_id] }
+      └─StreamDynamicFilter { predicate: ($expr1 > $expr2), output: [t.ts, t.a, $expr1, t._row_id] }
+        ├─StreamProject { exprs: [t.ts, t.a, AddWithTimeZone(t.ts, '01:00:00':Interval, 'UTC':Varchar) as $expr1, t._row_id] }
+        │ └─StreamTableScan { table: t, columns: [t.ts, t.a, t._row_id], stream_scan_type: ArrangementBackfill, stream_key: [t._row_id], pk: [_row_id], dist: UpstreamHashShard(t._row_id) }
+        └─StreamExchange { dist: Broadcast }
+          └─StreamProject { exprs: [DateTrunc('day':Varchar, SubtractWithTimeZone('2024-07-18 00:00:00+00:00':Timestamptz, ('2024-07-18 00:00:00+00:00':Timestamptz - now), 'UTC':Varchar), 'UTC':Varchar) as $expr2] }
+            └─StreamNow { output: [now] }
```
Comment (st1page) [inline src/frontend/planner_test/tests/testdata/output/temporal_filter.yaml:488]: we can add more test such as 
- `xxx < now_expression`
- `xxx = now_expression`
- a non-monotonic now expression such as 
```
create table t (ts timestamp with time zone, c text, i int);
explain create materialized view mv as select * from t where i > extract(hour from now()) ;
```

**Review by st1page — COMMENTED** [review]
Generally LGTM. 
As a user facing issue, could you please help to improve the doc about the temporal filter? And maybe we need to modify some out-of-date error message such as that in LogicalFilter::to_stream()

**Comment by stdrc** [comment]
> As a user facing issue, could you please help to improve the doc about the temporal filter? And maybe we need to modify some out-of-date error message such as that in LogicalFilter::to_stream()

Thanks for reminding. Error message is now updated. Will update document later.

File: src/frontend/planner_test/tests/testdata/output/temporal_filter.yaml (hunk 1 above)
Comment (stdrc) [reply src/frontend/planner_test/tests/testdata/output/temporal_filter.yaml:488]: We already have many `xxx < now_expression` and `xxx = now_expression`. Just added a non-mono expression test for failing case.

**Comment by stdrc** [comment]
> Will update document later.

Oh I just realized that we cannot update the docs for now, because `current` document is for v1.10, and this PR will be in 1.11. Will do it after we release 1.10...

**Review by kwannoel — APPROVED** [review]

**Review by chenzl25 — APPROVED** [review]

**Pull request merged**